File: src/auth/controller.py
from typing import Annotated, Any, Dict, Sequence

from litestar import Request, Response, post, status_codes
from litestar.controller import Controller
from litestar.enums import MediaType, RequestEncodingType
from litestar.params import Body
from litestar.security.jwt import Token

from src.auth import urls
from src.auth.schemas import LoginForm, RegistrationForm
from src.auth.service import AuthService
from src.person.models import Person


class AuthController(Controller):
    """Authentication controller"""

    tags: Sequence[str] = ["auth", "authentication"]

    @post(
        urls.REGISTRATION,
        status_code=status_codes.HTTP_201_CREATED,
        media_type=MediaType.JSON,
    )
    async def register(
        self,
        data: Annotated[
            RegistrationForm, Body(media_type=RequestEncodingType.URL_ENCODED)
        ],
        auth_service: AuthService,
    ) -> Dict:
        """Registration endpoint"""
        user = await auth_service.register(form=data)
        if user:
            return {"message": "User succesfully created"}
        else:
            return {"message": "User  was not created"}

    @post(
        urls.LOGIN,
        status_code=status_codes.HTTP_200_OK,
        media_type=MediaType.JSON,
    )
    async def login(
        self,
        data: Annotated[LoginForm, Body(media_type=RequestEncodingType.URL_ENCODED)],
        auth_service: AuthService,
    ) -> Response[Any]:
        """Login endpoint"""
        return await auth_service.login(data=data)

    @post(
        urls.LOGOUT,
        status_code=status_codes.HTTP_200_OK,
        media_type=MediaType.JSON,
    )
    async def logout(
        self,
        request: Request[Person, Token, Any],
        auth_service: AuthService,
    ) -> Response[Any]:
        """Logout endpoint"""
        return await auth_service.logout(token=request.auth)

File: internals/auth/controller.go
package auth

import (
	"net/http"

	"github.com/gin-gonic/gin"
)

type controller struct {
	as AuthService
}

func Controller(authService AuthService) *controller {
	return &controller{as: authService}
}

func (c *controller) Register(ctx *gin.Context) {
	var form RegistrationForm

	if err := ctx.ShouldBind(&form); err != nil {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request data"})
		return
	}

	if err := c.as.Register(ctx, &form); err != nil {
		if err.Error() == "user already exists" {
			ctx.JSON(http.StatusBadRequest, gin.H{"message": "User already exists"})
			return
		}
		ctx.JSON(http.StatusInternalServerError, gin.H{"message": "Registration failed"})
		return
	}

	ctx.JSON(http.StatusCreated, gin.H{"message": "Registration successful"})
}

func (c *controller) Login(ctx *gin.Context) {
	var form LoginForm
	if err := ctx.ShouldBind(&form); err != nil {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request data"})
		return
	}
	token, err := c.as.Authenticate(ctx, &form)
	if err != nil {
		ctx.JSON(http.StatusUnauthorized, gin.H{"message": "Login error"})
		return
	}

	ctx.Header("Authorization", "Bearer "+token)
	ctx.JSON(http.StatusOK, gin.H{
		"message": "Login successful",
	})
}

func (c *controller) Logout(ctx *gin.Context) {
	ctx.JSON(http.StatusCreated, gin.H{"message": "Logout successful"})
}
